x/crosschain/keeper/in_tx_tracker.go

```diff
@@ -0,0 +1,185 @@
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.InTxTrackerKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.InTxTracker
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		list = append(list, val)
+	}
+	return list
+}
+
+func (k Keeper) GetAllInTxTrackerForChain(ctx sdk.Context, chainID int64) (list []types.InTxTracker) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.InTxTrackerKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte(fmt.Sprintf("%d-", chainID)))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.InTxTracker
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		list = append(list, val)
+	}
+	return list
+}
+
+func (k Keeper) GetAllInTxTrackerForChainPaginated(ctx sdk.Context, chainID int64, pagination *query.PageRequest) (inTxTrackers []types.InTxTracker, pageRes *query.PageResponse, err error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(fmt.Sprintf("%s", types.InTxTrackerKeyPrefix)))
+	chainStore := prefix.NewStore(store, types.KeyPrefix(fmt.Sprintf("%d-", chainID)))
+	pageRes, err = query.Paginate(chainStore, pagination, func(key []byte, value []byte) error {
+		var inTxTracker types.InTxTracker
+		if err := k.cdc.Unmarshal(value, &inTxTracker); err != nil {
+			return err
+		}
+		inTxTrackers = append(inTxTrackers, inTxTracker)
+		return nil
+	})
+	return
+}
+
+func (k msgServer) AddToInTxTracker(goCtx context.Context, msg *types.MsgAddToInTxTracker) (*types.MsgAddToInTxTrackerResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	chain := k.zetaObserverKeeper.GetParams(ctx).GetChainFromChainID(msg.ChainId)
+	if chain == nil {
+		return nil, observerTypes.ErrSupportedChains
+	}
+
+	adminPolicyAccount := k.zetaObserverKeeper.GetParams(ctx).GetAdminPolicyAccount(observerTypes.Policy_Type_group1)
+	isAdmin := msg.Creator == adminPolicyAccount
+
+	isObserver, err := k.zetaObserverKeeper.IsAuthorized(ctx, msg.Creator, chain)
+	if err != nil {
+		ctx.Logger().Error("Error while checking if the account is an observer", err)
+	}
+	isProven := false
+	if msg.Proof != nil {
+		isProven, err = k.VerifyInTxTrackerProof(ctx, msg.Proof, msg.BlockHash, msg.TxIndex, msg.ChainId, msg.CoinType)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Sender needs to be either the admin policy account or an observer
+	if !(isAdmin || isObserver || isProven) {
+		return nil, errorsmod.Wrap(observerTypes.ErrNotAuthorized, fmt.Sprintf("Creator %s", msg.Creator))
+	}
+
+	k.Keeper.SetInTxTracker(ctx, types.InTxTracker{
+		ChainId:  msg.ChainId,
+		TxHash:   msg.TxHash,
+		CoinType: msg.CoinType,
+	})
+	return &types.MsgAddToInTxTrackerResponse{}, nil
+}
+
+func (k Keeper) VerifyInTxTrackerProof(ctx sdk.Context, proof *common.Proof, hash string, txIndex int64, chainID int64, coinType common.CoinType) (bool, error) {
```

Comment:
Do we need a verify proof function specific for `InTxTrackerProof`? I would think that we can have a single verify function used by other tx to check in the future like outbound, etc..

Reply:
The verification has two components to it , 
1 ) common for all msgs
2) specific for a msg-type 

Ideally, we should have two verification functions, for the two things mentioned above, which would need a minor refactor. I would do that in a separate PR  
https://github.com/zeta-chain/node/issues/1254